File: silicon_fury/fighter.py
"""Fighter entity — articulated limbs, aerial combat, explosive hits."""

from __future__ import annotations

import math
import random
from typing import TYPE_CHECKING, List, Optional

import pygame

from silicon_fury.body import draw_fighter_body
from silicon_fury.characters import Character
from silicon_fury.config import GRAVITY, GROUND_Y, WIDTH

if TYPE_CHECKING:
    from silicon_fury.effects import EffectWorld


class Fighter:
    def __init__(self, char: Character, x: float, facing: int = 1, is_cpu: bool = False):
        self.char = char
        self.x = x
        self.y = float(GROUND_Y)
        self.vx = 0.0
        self.vy = 0.0
        self.facing = facing
        self.is_cpu = is_cpu

        self.max_hp = 1000 + char.hp * 8
        self.hp = float(self.max_hp)
        self.meter = 0.0
        self.max_meter = 100.0

        self.state = "idle"
        self.state_t = 0
        self.state_max = 0
        self.hitstun = 0
        self.attack_hit = False
        self.on_ground = True
        self.combo = 0
        self.round_wins = 0
        self.flash = 0
        self.walk_phase = random.random() * 10
        self.air_jumps = 0
        self.dash_t = 0

        self.move_speed = 4.6 + char.speed * 0.05
        self.jump_power = 16.5 + char.speed * 0.03
        self.punch_dmg = 34 + char.power * 0.26
        self.kick_dmg = 48 + char.power * 0.32
        self.special_dmg = 125 + char.special * 0.65
        self.defense_factor = 1.0 - (char.defense / 220.0)
        self.reach_px = 78 + char.reach * 0.6

        # local sparks (also mirrored into EffectWorld on hits)
        self.sparks: List[dict] = []

    def set_state(self, state: str, frames: int = 12) -> None:
        self.state = state
        self.state_t = frames
        self.state_max = frames
        if state in {"punch", "kick", "air_kick", "special"}:
            self.attack_hit = False

    def can_act(self) -> bool:
        return self.hp > 0 and self.hitstun <= 0 and self.state not in {"ko", "special"}

    def _attacking(self) -> bool:
        return self.state in {"punch", "kick", "air_kick", "special"}

    def move(self, direction: int) -> None:
        if self.hp <= 0 or self.hitstun > 0 or self.state in {"ko", "special", "block"}:
            return
        # Seamless: allow movement during late attack recovery / air
        if self._attacking() and self.state_t > self.state_max * 0.35:
            return
        speed = self.move_speed if self.on_ground else self.move_speed * 0.85
        self.vx = direction * speed
        self.facing = 1 if direction > 0 else -1
        if self.on_ground and not self._attacking():
            self.set_state("walk", 10)
            self.walk_phase += 0.35

    def dash(self, direction: int) -> None:
        if not self.can_act() or self.dash_t > 0:
            return
        self.facing = 1 if direction > 0 else -1
        self.vx = self.facing * (self.move_speed * 2.4)
        self.dash_t = 14
        if self.on_ground:
            self.set_state("walk", 10)

    def jump(self) -> None:
        if self.hp <= 0 or self.hitstun > 0 or self.state in {"ko", "special"}:
            return
        if self.on_ground:
            self.vy = -self.jump_power
            self.on_ground = False
            self.air_jumps = 1
            self.set_state("jump", 28)
        elif self.air_jumps > 0 and self.state not in {"air_kick"}:
            # short double-hop for aerial freedom
            self.vy = -self.jump_power * 0.72
            self.air_jumps -= 1
            self.set_state("jump", 20)

    def block(self, holding: bool) -> None:
        if self.hp <= 0 or self.hitstun > 0:
            return
        if holding and self.on_ground and self.state not in {"punch", "kick", "air_kick", "special"}:
            self.set_state("block", 4)
            self.vx *= 0.2

    def punch(self) -> None:
        if self.hp <= 0 or self.hitstun > 0 or self.state in {"ko", "special"}:
            return
        if self.state in {"punch", "kick", "air_kick"} and self.state_t > 6:
            return
        if not self.on_ground:
            # air punch / jab
            self.set_state("punch", 14)
            self.vx += self.facing * 1.5
            return
        self.set_state("punch", 14)
        self.vx = self.facing * 3.2

    def kick(self) -> None:
        if self.hp <= 0 or self.hitstun > 0 or self.state in {"ko", "special"}:
            return
        if self.state in {"punch", "kick", "air_kick"} and self.state_t > 6:
            return
        if not self.on_ground:
            self.set_state("air_kick", 18)
            self.vx = self.facing * 3.5
            self.vy = min(self.vy, -2.5)  # hang for the kick
            return
        self.set_state("kick", 18)
        self.vx = self.facing * 3.8

    def special(self, fx: Optional["EffectWorld"] = None) -> bool:
        if self.meter < 55 or self.hp <= 0 or self.hitstun > 0 or self.state == "ko":
            return False
        if self.state == "special":
            return False
        self.meter -= 55
        self.set_state("special", 38)
        self.vx = self.facing * (4.2 + self.char.speed * 0.025)
        if not self.on_ground:
            self.vy = -3
        if fx:
            fx.explosion(self.x + self.facing * 40, self.y - 110, self.char.accent, 0.85)
            fx.fire_burst(self.x, self.y - 40, 0.7)
        for _ in range(14):
            self.sparks.append(
                {
                    "x": self.x,
                    "y": self.y - 120,
                    "vx": self.facing * random.uniform(2, 9),
                    "vy": random.uniform(-5, 2),
                    "life": random.randint(12, 26),
                    "color": self.char.accent,
                }
            )
        return True

    def take_hit(
        self,
        dmg: float,
        knock: float,
        attacker_facing: int,
        fx: Optional["EffectWorld"] = None,
        kind: str = "punch",
    ) -> None:
        blocked = self.state == "block"
        impact_x = self.x - attacker_facing * 12
        impact_y = self.y - (140 if kind == "punch" else 90)
        if blocked:
            dmg *= 0.28
            knock *= 0.18
            self.meter = min(self.max_meter, self.meter + 6)
            if fx:
                fx.hit_sparks(impact_x, impact_y, attacker_facing, (255, 220, 120))
        else:
            self.hitstun = 10 if kind != "special" else 16
            self.set_state("hit", self.hitstun)
            self.flash = 8
            self.combo = 0
            if fx:
                power = 1.0 if kind == "punch" else (1.35 if kind == "kick" or kind == "air_kick" else 2.0)
                fx.blood_burst(impact_x, impact_y, attacker_facing, power)
                fx.hit_sparks(impact_x, impact_y, attacker_facing, self.char.accent)
                if kind in {"kick", "air_kick"}:
                    fx.fire_burst(impact_x, impact_y, 0.45)
                if kind == "special":
                    fx.explosion(impact_x, impact_y, self.char.accent, 1.3)
                    fx.blood_burst(impact_x, impact_y - 20, attacker_facing, 1.8)
        dmg *= self.defense_factor
        self.hp = max(0, self.hp - dmg)
        self.vx = attacker_facing * knock
        self.vy = -5.5 if not blocked else -1.5
        if kind == "air_kick":
            self.vy = -7.5
        self.on_ground = False
        if self.hp <= 0:
            self.hp = 0
            self.set_state("ko", 140)
            self.hitstun = 140
            if fx:
                fx.blood_burst(self.x, self.y - 100, attacker_facing, 2.4)
                fx.explosion(self.x, self.y - 80, (255, 60, 40), 1.6)
                fx.fire_burst(self.x, self.y - 20, 1.2)

    def gain_meter(self, amount: float) -> None:
        self.meter = min(self.max_meter, self.meter + amount)

    def attack_box(self) -> Optional[pygame.Rect]:
        def box(x_off: float, y: float, w: int, h: int) -> pygame.Rect:
            x = int(self.x + self.facing * x_off)
            if self.facing < 0:
                x -= w
            return pygame.Rect(x, int(y), w, h)

        prog = 1.0 - (self.state_t / max(1, self.state_max))
        if self.state == "punch" and 0.22 <= prog <= 0.72:
            return box(48, self.y - 145, int(self.reach_px), 50)
        if self.state == "kick" and 0.28 <= prog <= 0.78:
            return box(42, self.y - 100, int(self.reach_px + 36), 56)
        if self.state == "air_kick" and 0.25 <= prog <= 0.8:
            return box(40, self.y - 120, int(self.reach_px + 40), 60)
        if self.state == "special" and 0.18 <= prog <= 0.85:
            return box(24, self.y - 170, int(self.reach_px + 100 + self.char.special * 0.4), 120)
        return None

    def body_box(self) -> pygame.Rect:
        return pygame.Rect(int(self.x - 36), int(self.y - 200), 72, 200)

    def update(self) -> None:
        if self.hitstun > 0:
            self.hitstun -= 1
        if self.dash_t > 0:
            self.dash_t -= 1
        if self.state_t > 0:
            self.state_t -= 1
            if self.state_t <= 0 and self.state != "ko":
                if not self.on_ground:
                    self.state = "jump"
                    self.state_t = 8
                    self.state_max = 8
                else:
                    self.state = "idle"
                    self.state_max = 0

        if self.flash > 0:
            self.flash -= 1

        if self.state == "walk":
            self.walk_phase += 0.42
        elif self.state == "idle" and self.on_ground:
            self.walk_phase += 0.08

        self.vy += GRAVITY
        self.x += self.vx
        self.y += self.vy
        # less sticky friction so movement feels free
        self.vx *= 0.90 if self.on_ground else 0.96

        if self.y >= GROUND_Y:
            self.y = GROUND_Y
            self.vy = 0
            if not self.on_ground and self.state not in {"ko", "hit"}:
                if self.state in {"jump", "air_kick"}:
                    self.state = "idle"
            self.on_ground = True
            self.air_jumps = 1
        else:
            self.on_ground = False

        self.x = max(70, min(WIDTH - 70, self.x))
        if self.state == "idle":
            self.gain_meter(0.12)

        for s in self.sparks:
            s["x"] += s["vx"]
            s["y"] += s["vy"]
            s["life"] -= 1
        self.sparks = [s for s in self.sparks if s["life"] > 0]

    def draw(self, surf: pygame.Surface) -> None:
        for s in self.sparks:
            pygame.draw.circle(surf, s["color"], (int(s["x"]), int(s["y"])), 3)

        # Special aura — ring only (no filled blob that reads as a dark box)
        if self.state == "special":
            pulse = self.state_max - self.state_t
            radius = 70 + pulse * 4
            aura = pygame.Surface((radius * 2, radius * 2), pygame.SRCALPHA)
            pygame.draw.circle(aura, (*self.char.accent, 110), (radius, radius), radius, 6)
            pygame.draw.circle(aura, (*self.char.primary, 70), (radius, radius), int(radius * 0.72), 3)
            surf.blit(aura, (self.x - radius, self.y - 130 - radius // 2))

        draw_fighter_body(
            surf,
            self.char,
            self.x,
            self.y,
            self.facing,
            self.state,
            self.state_t,
            self.state_max,
            self.on_ground,
            self.flash,
            self.walk_phase,
        )

        if self.state == "special":
            font = pygame.font.SysFont("Impact", 30)
            label = font.render(self.char.special_name, True, self.char.accent)
            outline = font.render(self.char.special_name, True, (0, 0, 0))
            lx = self.x - label.get_width() / 2
            ly = self.y - 290
            surf.blit(outline, (lx + 2, ly + 2))
            surf.blit(label, (lx, ly))

File: web/src/fighter.ts
import { CLIP_DURATION, sampleClip, type ClipName } from "./anim/clips";
import { drawSkeleton } from "./anim/skeleton";
import type { FighterKit } from "./types";
import { GROUND_Y, GRAVITY, clamp, type Pose } from "./types";
import type { EffectWorld } from "./effects";

export type FighterState =
  | "idle"
  | "walk"
  | "run"
  | "dash"
  | "jump"
  | "fall"
  | "crouch"
  | "block"
  | "punch"
  | "punch2"
  | "kick"
  | "airKick"
  | "special"
  | "hit"
  | "ko";

const ATTACKS = new Set<FighterState>(["punch", "punch2", "kick", "airKick", "special"]);

export class Fighter {
  kit: FighterKit;
  x: number;
  y: number;
  vx = 0;
  vy = 0;
  facing: 1 | -1;
  isCpu: boolean;

  maxHp: number;
  hp: number;
  meter = 0;
  maxMeter = 100;
  roundWins = 0;

  state: FighterState = "idle";
  stateT = 0;
  stateDur = 0;
  animT = 0;
  comboStep = 0;
  comboWindow = 0;
  hitstun = 0;
  attackHit = false;
  onGround = true;
  airJumps = 1;
  dashCd = 0;
  flash = 0;
  invuln = 0;

  moveSpeed: number;
  jumpPower: number;
  punchDmg: number;
  kickDmg: number;
  specialDmg: number;
  defenseFactor: number;
  reach: number;

  constructor(kit: FighterKit, x: number, facing: 1 | -1, isCpu = false) {
    this.kit = kit;
    this.x = x;
    this.y = GROUND_Y;
    this.facing = facing;
    this.isCpu = isCpu;
    this.maxHp = 1000 + kit.hp * 8;
    this.hp = this.maxHp;
    this.moveSpeed = 280 + kit.speed * 3.2;
    this.jumpPower = 780 + kit.speed * 2.2;
    this.punchDmg = 38 + kit.power * 0.28;
    this.kickDmg = 52 + kit.power * 0.34;
    this.specialDmg = 130 + kit.special * 0.7;
    this.defenseFactor = 1 - kit.defense / 230;
    this.reach = 78 + kit.power * 0.15;
  }

  get pose(): Pose {
    const clip = this.state as ClipName;
    const loop =
      this.state === "idle" ||
      this.state === "walk" ||
      this.state === "run" ||
      this.state === "fall" ||
      this.state === "crouch" ||
      this.state === "block";
    const dur = CLIP_DURATION[clip] || 0.4;
    if (loop) return sampleClip(clip, this.animT / dur, true);
    const t = this.stateDur > 0 ? 1 - this.stateT / this.stateDur : 0;
    return sampleClip(clip, t, false);
  }

  canAct(): boolean {
    return this.hp > 0 && this.hitstun <= 0 && this.state !== "ko" && this.state !== "special";
  }

  private setState(state: FighterState, dur?: number): void {
    this.state = state;
    this.stateDur = dur ?? CLIP_DURATION[state as ClipName] ?? 0.3;
    this.stateT = this.stateDur;
    this.animT = 0;
    if (ATTACKS.has(state)) this.attackHit = false;
  }

  move(dir: -1 | 1, sprint = false): void {
    if (this.hp <= 0 || this.hitstun > 0 || this.state === "ko" || this.state === "special") return;
    if (ATTACKS.has(this.state) && this.stateT > this.stateDur * 0.35) return;
    if (this.state === "block" || this.state === "crouch") return;

    const spd = (sprint ? this.moveSpeed * 1.45 : this.moveSpeed) * (this.onGround ? 1 : 0.85);
    this.vx = dir * spd;
    this.facing = dir;
    if (this.onGround && !ATTACKS.has(this.state) && this.state !== "dash") {
      this.setState(sprint ? "run" : "walk");
    }
  }

  dash(dir: -1 | 1): void {
    if (!this.canAct() || this.dashCd > 0 || !this.onGround) return;
    this.facing = dir;
    this.vx = dir * this.moveSpeed * 2.8;
    this.dashCd = 0.45;
    this.setState("dash", CLIP_DURATION.dash);
    this.invuln = 0.08;
  }

  jump(): void {
    if (this.hp <= 0 || this.hitstun > 0 || this.state === "ko" || this.state === "special") return;
    if (this.onGround) {
      this.vy = -this.jumpPower;
      this.onGround = false;
      this.airJumps = 1;
      this.setState("jump");
    } else if (this.airJumps > 0 && this.state !== "airKick") {
      this.vy = -this.jumpPower * 0.72;
      this.airJumps -= 1;
      this.setState("jump", 0.35);
    }
  }

  crouch(holding: boolean): void {
    if (this.hp <= 0 || this.hitstun > 0 || !this.onGround) return;
    if (holding && !ATTACKS.has(this.state) && this.state !== "special") {
      this.setState("crouch");
      this.vx *= 0.2;
    }
  }

  block(holding: boolean): void {
    if (this.hp <= 0 || this.hitstun > 0) return;
    if (holding && this.onGround && !ATTACKS.has(this.state) && this.state !== "special") {
      this.setState("block");
      this.vx *= 0.15;
    }
  }

  punch(): void {
    if (this.hp <= 0 || this.hitstun > 0 || this.state === "ko" || this.state === "special") return;
    if (ATTACKS.has(this.state) && this.stateT > 0.08) return;
    if (!this.onGround) {
      this.setState("punch", 0.28);
      this.vx += this.facing * 40;
      return;
    }
    if (this.comboWindow > 0 && this.comboStep === 1) {
      this.comboStep = 2;
      this.setState("punch2");
      this.vx = this.facing * 160;
      return;
    }
    this.comboStep = 1;
    this.comboWindow = 0.35;
    this.setState("punch");
    this.vx = this.facing * 140;
  }

  kick(): void {
    if (this.hp <= 0 || this.hitstun > 0 || this.state === "ko" || this.state === "special") return;
    if (ATTACKS.has(this.state) && this.stateT > 0.08) return;
    if (!this.onGround) {
      this.setState("airKick");
      this.vx = this.facing * 120;
      this.vy = Math.min(this.vy, -80);
      return;
    }
    this.setState("kick");
    this.vx = this.facing * 180;
    this.comboStep = 0;
  }

  special(fx?: EffectWorld): boolean {
    if (this.meter < 55 || this.hp <= 0 || this.hitstun > 0 || this.state === "ko") return false;
    if (this.state === "special") return false;
    this.meter -= 55;
    this.setState("special");
    this.vx = this.facing * (220 + this.kit.speed);
    if (!this.onGround) this.vy = -120;
    fx?.explosion(this.x + this.facing * 50, this.y - 120, this.kit.accent, 0.9);
    return true;
  }

  takeHit(
    dmg: number,
    knock: number,
    atkFacing: 1 | -1,
    kind: string,
    fx?: EffectWorld,
  ): void {
    const blocked = this.state === "block";
    const ix = this.x - atkFacing * 14;
    const iy = this.y - (kind === "punch" || kind === "punch2" ? 145 : 95);
    if (blocked) {
      dmg *= 0.28;
      knock *= 0.18;
      this.meter = Math.min(this.maxMeter, this.meter + 6);
      fx?.sparks(ix, iy, atkFacing, [255, 220, 120]);
    } else {
      this.hitstun = kind === "special" ? 0.35 : 0.22;
      this.setState("hit", this.hitstun);
      this.flash = 0.18;
      const power = kind === "special" ? 2 : kind.includes("kick") ? 1.35 : 1;
      fx?.blood(ix, iy, atkFacing, power);
      fx?.sparks(ix, iy, atkFacing, this.kit.accent);
      if (kind.includes("kick") || kind === "special") fx?.fire(ix, iy, kind === "special" ? 1.1 : 0.45);
      if (kind === "special") fx?.explosion(ix, iy, this.kit.accent, 1.35);
    }
    this.hp = Math.max(0, this.hp - dmg * this.defenseFactor);
    this.vx = atkFacing * knock * 60;
    this.vy = blocked ? -80 : kind === "airKick" ? -420 : -280;
    this.onGround = false;
    if (this.hp <= 0) {
      this.hp = 0;
      this.setState("ko", CLIP_DURATION.ko);
      this.hitstun = CLIP_DURATION.ko;
      fx?.blood(this.x, this.y - 100, atkFacing, 2.4);
      fx?.explosion(this.x, this.y - 80, [255, 60, 40], 1.5);
    }
  }

  attackBox(): { x: number; y: number; w: number; h: number } | null {
    if (!ATTACKS.has(this.state) || this.attackHit) return null;
    const prog = 1 - this.stateT / Math.max(1e-3, this.stateDur);
    const box = (xOff: number, y: number, w: number, h: number) => {
      let x = this.x + this.facing * xOff;
      if (this.facing < 0) x -= w;
      return { x, y, w, h };
    };
    if (this.state === "punch" && prog > 0.2 && prog < 0.55) return box(48, this.y - 150, this.reach, 48);
    if (this.state === "punch2" && prog > 0.22 && prog < 0.58) return box(52, this.y - 148, this.reach + 8, 50);
    if (this.state === "kick" && prog > 0.25 && prog < 0.62) return box(46, this.y - 105, this.reach + 36, 54);
    if (this.state === "airKick" && prog > 0.2 && prog < 0.7) return box(44, this.y - 120, this.reach + 40, 56);
    if (this.state === "special" && prog > 0.28 && prog < 0.75)
      return box(30, this.y - 170, this.reach + 110 + this.kit.special * 0.35, 120);
    return null;
  }

  bodyBox(): { x: number; y: number; w: number; h: number } {
    const crouch = this.state === "crouch" || this.state === "block";
    const h = crouch ? 140 : 200;
    return { x: this.x - 34, y: this.y - h, w: 68, h };
  }

  update(dt: number): void {
    if (this.hitstun > 0) this.hitstun -= dt;
    if (this.dashCd > 0) this.dashCd -= dt;
    if (this.flash > 0) this.flash -= dt;
    if (this.invuln > 0) this.invuln -= dt;
    if (this.comboWindow > 0) {
      this.comboWindow -= dt;
      if (this.comboWindow <= 0) this.comboStep = 0;
    }

    this.animT += dt;
    const looping =
      this.state === "idle" ||
      this.state === "walk" ||
      this.state === "run" ||
      this.state === "fall" ||
      this.state === "crouch" ||
      this.state === "block";

    if (!looping && this.stateT > 0) {
      this.stateT -= dt;
      if (this.stateT <= 0 && this.state !== "ko") {
        if (!this.onGround) this.setState("fall");
        else this.setState("idle");
      }
    } else if (looping) {
      // Keep locomotion / hold poses looping — never snap mid-stride.
      this.stateT = this.stateDur;
    }

    this.vy += GRAVITY * dt;
    this.x += this.vx * dt;
    this.y += this.vy * dt;
    this.vx *= this.onGround ? 0.82 : 0.96;

    if (this.y >= GROUND_Y) {
      this.y = GROUND_Y;
      this.vy = 0;
      if (!this.onGround && this.state !== "ko" && this.state !== "hit") {
        if (this.state === "jump" || this.state === "fall" || this.state === "airKick") this.setState("idle");
      }
      this.onGround = true;
      this.airJumps = 1;
    } else {
      this.onGround = false;
      if (this.state === "idle" || this.state === "walk" || this.state === "run") this.setState("fall");
    }

    this.x = clamp(this.x, 70, 1210);
    if (this.state === "idle") this.meter = Math.min(this.maxMeter, this.meter + 8 * dt);

    // Stop walk/run when nearly still
    if (
      this.onGround &&
      (this.state === "walk" || this.state === "run") &&
      Math.abs(this.vx) < 25
    ) {
      this.setState("idle");
    }
  }

  draw(ctx: CanvasRenderingContext2D): void {
    const attackFlash =
      ATTACKS.has(this.state) && this.stateT / this.stateDur < 0.65 && this.stateT / this.stateDur > 0.2
        ? 0.7
        : 0;
    drawSkeleton(ctx, this.kit, this.x, this.y, this.facing, this.pose, {
      flash: this.flash,
      airborne: !this.onGround,
      attackFlash,
    });

    if (this.state === "special") {
      ctx.save();
      ctx.font = "bold 28px Bebas Neue, Impact, sans-serif";
      ctx.textAlign = "center";
      ctx.fillStyle = "#000";
      ctx.fillText(this.kit.specialName, this.x + 2, this.y - 268);
      ctx.fillStyle = `rgb(${this.kit.accent.join(",")})`;
      ctx.fillText(this.kit.specialName, this.x, this.y - 270);
      ctx.restore();
    }
  }
}
